R/bs-theme-preset-brand.R

```diff
@@ -0,0 +1,830 @@
+# "brand" and "preset" interact in a few ways that need to be reconciled.
+# Ultimately, we want three layers:
+#
+# 1. The base Bootstrap styles, by version. E.g. Bootstrap 5
+# 2. The `preset` styles, which are one layer above and adjust the base
+#    Bootstrap with the preset defaults E.g. "shiny" (bslib) or "flatly" (Bootswatch)
+# 3. The `brand` styles, which can be seen as the final layer of default styles.
+#    These styles set Bootstrap variables to adjust the `preset` theme in the
+#    direction of the brand's color palette and typography.
+#
+# The `preset` and `version` can be specified in two ways with `brand`:
+#
+# 1. Directly via `bs_theme()`:
+#    `bs_theme(preset = "shiny", version = 5, brand = TRUE)`.
+#
+# 2. As part of the `brand` definition:
+#    `bs_theme(brand = list(defaults = list(shiny = list(theme = list(preset = "flatly")))))`
+#
+# So the order of operations is:
+#
+# 1. Read `brand`, if necessary.
+# 2. If `bs_theme()` provides guidance for `version` and `theme`, use those preferences.
+# 3. If not, inspect `brand` for guidance.
+# 4. Finally, use our own defaults.
+#
+# Importantly we need to separately read `brand` and then resolve the preset and
+# then resolve the brand bundle.
+
+bs_brand_bundle <- function(brand, version = version_default()) {
+  brand <- brand_resolve(brand)
+  
+  if (is.null(brand)) {
+    return()
+  }
+
+  brand_fonts <- brand_sass_fonts(brand)
+  brand_color_palette <- brand_sass_color_palette(brand)
+  brand_color <- brand_sass_color(brand)
+  brand_defaults <- brand_sass_defaults_bootstrap(brand)
+  brand_typography <- brand_sass_typography(brand)
+
+  sass_bundle(
+    "brand_base" = switch_version(
+      version,
+      five = sass_layer_file(
+        system_file("brand", "bs5", "_brand-yml.scss", package = "bslib")
+      ),
+      default = list()
+    ),
```

Comment:
Should we be throwing an error if version is less than 5?

Reply:
I'm not sure... we don't have a fully adapted branded theme for BS <= 4 but there are still a number of features of brand that are useful. It's also possible that a user could fill in their own implementation for BS <= 4 using only the tools already present in brand.yml. So erroring here feels too aggressive. I'll add a warning instead.